rust/chains/hyperlane-cosmos/src/trait_builder.rs

```diff
@@ -11,6 +16,46 @@ pub struct ConnectionConf {
     prefix: String,
     /// Canoncial Assets Denom
     canonical_asset: String,
+    /// The minimum gas price set by the cosmos-sdk validator
+    minimum_gas_price: RawCosmosAmount,
+}
+
+/// Untyped cosmos amount
+#[derive(serde::Serialize, serde::Deserialize, new, Clone, Debug)]
+pub struct RawCosmosAmount {
+    /// Coin denom (e.g. `untrn`)
+    pub denom: String,
+    /// Amount in the given denom
+    pub amount: String,
+}
+
+/// Typed cosmos amount
+#[derive(Clone, Debug)]
+pub struct CosmosAmount {
+    /// Coin denom (e.g. `untrn`)
+    pub denom: String,
+    /// Amount in the given denom
+    pub amount: U256,
+}
+
+impl TryFrom<RawCosmosAmount> for CosmosAmount {
+    type Error = ChainCommunicationError;
+    fn try_from(raw: RawCosmosAmount) -> Result<Self, ChainCommunicationError> {
+        // Converts to U256 by always rounding up.
+
+        // Remove the decimal part
+        let integer = raw
```

Comment:
Ah I realize now why you mentioned the rounding by removing the decimal part and adding 1, for some reason I didn't connect the dots on implications

It looks like the suggested Osmosis gas price is [`0.0025`](https://github.com/cosmos/chain-registry/blob/master/osmosis/chain.json#L22), which iiuc as written we would result in using a gas price of 1, so we'd be paying 400x for each tx than we need to. Similar for Neutron but not as bad as it's [`0.5`](https://github.com/cosmos/chain-registry/blob/master/neutron/chain.json#L20), so we'd be paying 2x

So I think we'll need to have the amount be able to be less than 1. Either as an f64 or maybe by specifying a numerator and denominator like how we do for `GasPaymentEnforcementPolicy::OnChainFeeQuoting`

I know for TxOutcome / TxCostEstimate we use `gas_price: U256` - I see our options here as:
1. just keep it as U256 and do something similar to what we do here where we ceil the f64 gas price for cosmos chains
2. we change TxOutcome to support < 1 gas prices, either via floats or int numerator / denominator, and then change the storage key for TxOutcomes in hyperlane db for these because this'll change the encoding of existing data. We could maybe even have an enum with a U256 variant and a f64 variant or something, but not sure what that'd give us tbh

I think 2 is probably the nicer solution longer

Reply:
TIL gas price is not something you're reimbursed for :/

Rather than have custom logic for the decimal part, I'm inclined to go for a crate that stores floats on the heap but abstracts the complexity away and can be arbitrarily large
https://crates.io/crates/bigdecimal